libsnark/zk_proof_systems/plonk/prover.tcc

```diff
@@ -1060,14 +1112,20 @@ plonk_proof<ppT> plonk_prover<ppT>::compute_proof(
         round_two_out,
         round_three_out,
         round_four_out,
-        srs);
+        srs,
+        hasher);
 
     // TODO: activate this part when we implement actual hashing of
     // communication transcripts
-#if 0    
+#if 0
     // u: multipoint evaluation challenge -- hash of transcript from
     // rounds 1,2,3,4,5
-    libff::Fr<ppT> u = transcript_hash.u;
+    const libff::Fr<ppT> u = hasher.get_hash();
```

Comment:
Why do we not need `u` here?  What will change when the actual hasher is implemented so that `u` is suddenly required?

Reply:
Good point. Strictly speaking, we don't need `u` to be computed here even if we are using a "real" hash function. Indeed, if you look at the paper [GWC19], pp.30, at the end of round 5 of the unrolled protocol the prover computes `u = H(transcript)` but does nothing with it. `u` is only used on the verifier side in step 9 ([GWC19], pp.31).

In my case, the reason to execute `hasher.get_hash()` at the end of the prover computation is in order to reach the maximum buffer length at which point the buffer is automatically cleared inside `get_hash()`.Upon closer look though, and prompted by your question, I now realize that there is no need to do that. The buffer can be cleared by the caller instead. The latter happens for example, when we want to re-use the same hasher object multiple times in the unit tests.

The above change is now applied and committed. In particular the computation of `u = hasher.get_hash();` is removed from the prover side.

Reply:
I'm not sure I fully follow this.  Let's discuss offline.

Reply:
Just to explain in the meantime, I think we should be hashing the entire transcript (or some derivative of it).  When the caller calls `get_hash` we could compute the hash, clear the buffer and write the new hash into the buffer, and this then allows use to call `get_hash` multiple times without adding more data (alternatively we coudl append it to the buffer).  But clearing the buffer, and allowing the calling code to do this, sounds like it could lead to problems.

W.r.t. this code, I'm fine with the prover calling `hasher.get_hash();` in order to stay in sync with verifier (as long as that's explained in a comment).  I actually think this is preferable to having having a `clear_buffer` on the transcript hasher interface (since, as discussed, it may be possible to implement a transcript_hasher without a buffer).

Reply:
I see your point, but I have a couple of clarifying questions/remarks. Let's discuss offline. It'll be faster too.

Reply:
`get_hash` may update the internal state of the transcript_hasher, depending on the implementation, therefore I think the prover and verifier must make exactly the same calls to both `add_element` *and* `get_hash`.

Let's make sure there is a comment so it's not confusing.

`libff::UNUSED(u)`

Reply:
Added back the computation of `u` in the prover. Now prover and verifier make the same number of calls to `get_hash()`.